### PR TITLE
Add atom timings to API response

### PR DESCRIPTION
**Pull request by josephharrington**
- Create a new Atom class that takes the place of our existing atoms
which were just strings
- Include both expected and actual time as fields in the Atom object,
and include these in the subjob API response.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @josephharrington has signed the CLA. Thanks for the pull request!

File: app/master/time_based_atom_grouper.py [hunk 1]
```diff
@@ -103,28 +103,27 @@ def groupings(self):
     def _coalesce_new_atoms_with_historic_times(self, new_atoms, old_atoms_with_times, project_directory):
```
Comment (tjlee0909) [inline app/master/time_based_atom_grouper.py]: I remember naming this method. I think it wasn't a great name then, and it's an even worse name now with this change.

Maybe we can rename it to _set_expected_times()?

File: app/master/time_based_atom_grouper.py [hunk 2]
```diff
@@ -103,28 +103,27 @@ def groupings(self):
...
         """
         Combine the historic timing data from previous atoms (for the same job) with the new atoms generated for this
-        job.
+        job. This will update the expected_time fields for the atom objects in new_atoms.
```
Comment (tjlee0909) [inline app/master/time_based_atom_grouper.py]: "Set the expected run time (new_atom.expected_time)  of each atom in new_atoms using historic timing data."

**Comment by tjlee0909** [comment]
Wow, I didn't think this change would involve any significant work. Good thing you nipped it in the bud before it got out of hand.

:+1:

**Comment by tjlee0909** [comment]
Btw, what do you plan on doing now with this added functionality of being able to view expected vs actual atom timing data. More dashboard stuff?

**Comment by josephharrington** [comment]
@tjlee0909 Yeah, maybe some dashboard stuff eventually. We could also just run one-off reports to get a sense of how much variability we are seeing in atom times.

File: app/master/time_based_atom_grouper.py (hunk 1 above)
Comment (josephharrington) [inline app/master/time_based_atom_grouper.py]: Will do.

File: app/master/time_based_atom_grouper.py (hunk 2 above)
Comment (josephharrington) [inline app/master/time_based_atom_grouper.py]: Yeah, nice. Will do.

File: app/master/time_based_atom_grouper.py [hunk 3]
```diff
@@ -74,18 +73,19 @@ def groupings(self):
...
-                = self._coalesce_new_atoms_with_historic_times(self._atoms, self._atom_time_map, self._project_directory)
+            total_estimated_runtime = self._coalesce_new_atoms_with_historic_times(
+                self._atoms, self._atom_time_map, self._project_directory)
         except _AtomTimingDataError:
             grouper = AtomGrouper(self._atoms, self._max_executors)
             return grouper.groupings()
 
         # 2). Sort them by time, and add them to an OrderedDict
-        sorted_atom_times_left = OrderedDict(sorted(new_atoms_with_times_list, key=itemgetter(1), reverse=True))
+        atoms_sorted_by_desc_time = sorted(self._atoms, key=lambda atom: atom.expected_time, reverse=True)
```
Comment (wjdhollow) [inline app/master/time_based_atom_grouper.py]: desc_time?

File: app/master/time_based_atom_grouper.py (hunk 1 above)
Comment (wjdhollow) [inline app/master/time_based_atom_grouper.py]: Ha. I was actually going to make the same comment.

**Comment by wjdhollow** [comment]
:+1: w/ minor comment

File: app/master/time_based_atom_grouper.py (hunk 3 above)
Comment (josephharrington) [inline app/master/time_based_atom_grouper.py]: Yeah "descending" -- I'll clarify the name.

**Pull request merged**